[Feature]Add cugraph local storage support to DGL

VibhuJawa opened the pull request:
## Description


This PR closes https://github.com/dmlc/dgl/issues/4672 by untegrating cugraph local storage into DGL. 


Accuracy Results on https://github.com/dmlc/dgl/tree/master/examples/pytorch/rgcn-hetero by using the same defaults mentioned there so **hyper parameter tuning ins not done.**  but we are still in a percentage point of the avg runs . 

| dataset | Avg(cugraph, 18 runs) | Avg(DGL, 5 runs) |
|---------|-----------------------|------------------|
| aifb    | 93.81                 | 94.44            |
| bgs     | 90.04                 | 91.04            |
| mutag   | 66.17                 | 67.37            |


This PR enables cugraph storage by using duck-typing by following the `Extending DGL Graph Sampling Pipeline
` RFC (See [link](https://hackmd.io/6PJixJj7Ssy19I-N9CV1Ag?view))  so all one needs to do is to switch the graph object from `DGLGraph` object to `CuGraphStorage` object. 

See below:
```diff

+from dgl.contrib.cugraph.convert import cugraph_storage_from_heterograph
+cugraph_g = cugraph_storage_from_heterograph(dgl_g)

sampler = dgl.dataloading.NeighborSampler(
        [15, 10, 5], prefetch_node_feats=['feat'], prefetch_labels=['label'])

train_dataloader = dgl.dataloading.DataLoader(
- dgl_g, 
+ cugraph_g,
train_idx, 
sampler, 
device=device, 
batch_size=1024,
shuffle=True,
drop_last=False, 
num_workers=0)

```


<!-- Brief description. Refer to the related issues if existed.
It'll be great if relevant reviewers can be assigned as well.-->

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

## Features Added
-  [x] cugraph Homogeneous Graph
-  [x]  cugraph Heterogeneous Graph
-  [x] Single GPU cugraph support
-  [x] Multi GPU cugraph support
-  [x] End to End Examples 
-  [x] from hetrogenous_graph

##  Tests: 
-  [x] Homogeneous Graph Sampling
-  [x] Heterogeneous Graph Sampling
-  [x] from hetrogenous_graph


## Out of scope:
<del> renumber_vertices_by_type added in API
<del>  to_hetrogenous_graph
<del> edge_subgraph
<del> subgraph

VibhuJawa committed:
Added cugraph_storage

VibhuJawa committed:
Added cugraph_storage

VibhuJawa committed:
Added Working Hetrogenous Workflow

VibhuJawa committed:
Added Working Hetrogenous Workflow

VibhuJawa committed:
cugraph_storage multi-gpu updates

VibhuJawa committed:
cugraph_storage multi-gpu updates

VibhuJawa committed:
heterogeneous_cugraph_support

VibhuJawa committed:
heterogeneous_cugraph_support fixes

VibhuJawa committed:
added test for cugraph storage

VibhuJawa committed:
added check for canonical_etype

VibhuJawa committed:
heterogeneous_cugraph_support

VibhuJawa committed:
Added cugraph_storage

VibhuJawa committed:
Added working DGL HetroGraph Conversion

VibhuJawa commented on tests/cugraph/test_from_dgl_hetrograph.py:
```diff
@@ -0,0 +1,200 @@
+        create_test_heterograph4,
+    ]
+    for graph_f in graph_fs:
+        g = graph_f(th.int32)
+        gs = cugraph_storage_from_heterograph(g)
+        assert_same_num_nodes(gs, g)
+        try:
+            # TODO: Raise an upstream issue for this
+            # to triage first time failure
+            assert_same_node_feats(gs, g)
+        except ValueError:
+            assert_same_node_feats(gs, g)
+        assert_same_num_edges_can_etypes(gs, g)
+        assert_same_num_edges_etypes(gs, g)
+        assert_same_edge_feats(gs, g)
```
Apart from sampling everything is tested here.

VibhuJawa committed:
Added Tests

VibhuJawa committed:
Add newline to __init__.py

VibhuJawa committed:
Update __init__.py

VibhuJawa committed:
Works on MAG dataset

VibhuJawa committed:
passes tests

VibhuJawa committed:
Added cugraph_conversion_utils

VibhuJawa committed:
Added working Pytests

VibhuJawa committed:
Merge branch 'dmlc_upstream_master' into heterogeneous_cugraph_support

VibhuJawa commented on tests/cugraph/test_from_dgl_hetrograph.py:
```diff
@@ -0,0 +1,201 @@
+            try:
+                # TODO: Raise an upstream issue for this
+                # to triage first time failure
+                assert_same_node_feats(gs, g)
```
Fails due to a PG issue that i need to find a minimal example for.

VibhuJawa replied:
Will go away after this is merged: https://github.com/rapidsai/cugraph/pull/2804

alexbarghi-nv commented:
Looks fine to me.

VibhuJawa committed:
Added device property

VibhuJawa committed:
Merge branch 'heterogeneous_cugraph_support' of https://github.com/VibhuJawa/dgl into heterogeneous_cugraph_support

VibhuJawa committed:
Added example

VibhuJawa commented on examples/cugraph-pytorch/cugraph-local/rgcn-hetero/model.py:
```diff
@@ -0,0 +1,569 @@
+"""RGCN layer implementation"""
+# File same as: https://github.com/dmlc/dgl/blob/master/examples/pytorch/rgcn-hetero/model.py
+from collections import defaultdict
```
No need to review as same as https://github.com/dmlc/dgl/blob/master/examples/pytorch/rgcn-hetero/model.py

VibhuJawa commented on examples/cugraph-pytorch/cugraph-local/rgcn-hetero/entity_classify_mb.py:
```diff
@@ -0,0 +1,275 @@
+    else:
+        val_idx = train_idx
+    
+    #Only change done
+    if args.use_cugraph_storage:
+        g = cugraph_storage_from_heterograph(g)
+
```
Only change done from `cugraph_storage_from_heterograph`

wangxiaoyunNV replied:
ok

VibhuJawa committed:
Minor code cleanup around nodes

VibhuJawa committed:
Minor style-fixes

VibhuJawa committed:
Removed try expect block as upstream PG error is now fixed

VibhuJawa committed:
Added homogenous sampling test

VibhuJawa committed:
Removed unused imports

VibhuJawa commented on examples/cugraph-pytorch/cugraph-local/rgcn-hetero/entity_classify_mb.py:
```diff
@@ -0,0 +1,275 @@
+# Only change is:
+# from dgl.contrib.cugraph.convert import cugraph_storage_from_heterograph
+# g = cugraph_storage_from_heterograph(g)
+
```
Please see above comments as that  is the only change from upstream :-)

VibhuJawa committed:
Added doc string

VibhuJawa committed:
Added num_nodes_dict as a compulsary element for cleaner API

VibhuJawa committed:
Update README.MD

VibhuJawa committed:
Merge branch 'master' into heterogeneous_cugraph_support

VibhuJawa committed:
Update test_basics.py

VibhuJawa committed:
Update test_basics.py

VibhuJawa committed:
Cleaned up docstring example and added edge storage and node_stroage

VibhuJawa committed:
Merge branch 'heterogeneous_cugraph_support' of https://github.com/VibhuJawa/dgl into heterogeneous_cugraph_support

VibhuJawa committed:
Added tests

VibhuJawa committed:
fix import errors

VibhuJawa committed:
fix import errors

wangxiaoyunNV commented on python/dgl/contrib/cugraph/convert.py:
```diff
@@ -0,0 +1,21 @@
+import dgl
+from .utils.cugraph_conversion_utils import (
```
When from DGL to cugraph, did we ignore the dangling nodes?

VibhuJawa replied:
Dangling nodes with associated features will be handled , Can you give an example of a case you think is not handled , cause i think they should all be handled .

wangxiaoyunNV commented on python/dgl/contrib/cugraph/utils/cugraph_conversion_utils.py:
```diff
@@ -0,0 +1,178 @@
+    return gs
+
+
+def add_nodes_from_dgl_heteroGraph(
```
Did we also consider adding the dangling nodes here?

wangxiaoyunNV reviewed (commented):
In the conversion function, will the dangling nodes also be taken care of? We do not what to ignore them.

VibhuJawa committed:
Ensuring correct index-types

VibhuJawa committed:
Merge branch 'master' into heterogeneous_cugraph_support

VibhuJawa committed:
Merge branch 'master' into heterogeneous_cugraph_support

VibhuJawa committed:
Mofied existing example directly

VibhuJawa committed:
Merge branch 'heterogeneous_cugraph_support' of https://github.com/VibhuJawa/dgl into heterogeneous_cugraph_support

VibhuJawa committed:
Update entity_classify_mb.py

VibhuJawa committed:
Merge branch 'master' into heterogeneous_cugraph_support

kylasa committed:
Updated the key to retrieve correct rank of a process (#4756)

Merging this PR to the master branch

VibhuJawa committed:
Update cugraph_nightly_torch (#4757)

peizhou001 committed:
[Dist] Enable save and load for Distributed Optimizer (#4752)

* add save/load for distributed  optimizer

Co-authored-by: Ubuntu <ubuntu@ip-172-31-16-19.ap-northeast-1.compute.internal>

Kh4L committed:
[Dist][Test] Add tests for multi-node DistGraph (#4396)

Signed-off-by: Serge Panev <spanev@nvidia.com>

Signed-off-by: Serge Panev <spanev@nvidia.com>

VibhuJawa committed:
Merge branch 'master' into heterogeneous_cugraph_support

ZhenyuLU-Heliodore committed:
[NN] Add DegreeEncoder for graph transformer (#4742)

* Add files via upload

This file will contain several utility modules for Graph Transformer. DegreeEncoder has been implemented in this file now.

* Update graph_transformer.py

* Update nn-pytorch.rst

Add graph_transformer.DegreeEncoder in Utility Modules

* Update test_nn.py

* Update __init__.py

* Update graph_transformer.py

* Update test_nn.py

* Update graph_transformer.py

fix unused import

* Update graph_transformer.py

add module doc-string

* Update graph_transformer.py

* Update graph_transformer.py

Co-authored-by: rudongyu <ru_dongyu@outlook.com>

frozenbugs committed:
[Misc] clang-format auto fix. (#4800)

* clang-format

* manul

* manul

* manual

Co-authored-by: Steve <ubuntu@ip-172-31-34-29.ap-northeast-1.compute.internal>

frozenbugs committed:
[Misc] clang-format auto fix. (#4803)

* [Misc] clang-format auto fix.

* manual

Co-authored-by: Steve <ubuntu@ip-172-31-34-29.ap-northeast-1.compute.internal>

VibhuJawa committed:
Merge branch 'master' into heterogeneous_cugraph_support
